### PR TITLE
[SPARK-25238][PYTHON] lint-python: Upgrade pycodestyle to v2.4.0

### PR DESCRIPTION
**Pull request by cclauss**
See https://pycodestyle.readthedocs.io/en/latest/developer.html#changes for changes made in this release.

## What changes were proposed in this pull request?

Upgrade pycodestyle to v2.4.0

## How was this patch tested?

__pycodestyle__

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
Eh, @cclauss, mind filing a JIRA please?

**Comment by HyukjinKwon** [comment]
cc @rekhajoshm FYI

**Comment by cclauss** [comment]
https://issues.apache.org/jira/browse/SPARK-25238

**Comment by HyukjinKwon** [comment]
Mind fixing the PR title to `[SPARK-25238][PYTHON] lint-python: Upgrade pycodestyle to v2.4.0` so that it links the PR to JIRA?

**Comment by HyukjinKwon** [comment]
I think we can simply skip, for example, `W504 line break after binary operator` per PEP8:

> In Python code, it is permissible to break before or after a binary operator, as long as the convention is consistent locally.

**Comment by srowen** [comment]
Interesting, errors like:

```
./dev/run-tests.py:172:7: W605 invalid escape sequence '\d'
```

on lines like:

```
match = re.search('(\d+)\.(\d+)\.(\d+)', raw_version_str)
```

How does the above actually work? I'd think `\d` just produces `d` in the pattern as backslash is also a Python escape, and this isn't a raw string. Does Python not interpret `\d` in this case because `d` isn't escaped in Python? https://docs.python.org/3/library/re.html

Might be worth heeding the warning and making all of these instances raw strings, to avoid the warning and follow the advice in the link above.

**Comment by rekhajoshm** [comment]
thanks for the fyi @HyukjinKwon , ack.

**Comment by srowen** [comment]
What do you think @cclauss ?

**Comment by srowen** [comment]
@holdenk what do you think about the change I proposed above? I didn't know enough to know whether this was the right fix

**Comment by cclauss** [comment]
@srowen Could you please provide some test cases in Python 2 and Python 3 to prove that the changes that you propose will be properly interpreted without side effects in both?

**Comment by holdenk** [comment]
I'm a little slammed this week until Friday, I'll try and cycle back then.

**Comment by srowen** [comment]
I can really answer my own question, I'm just being lazy. This seems to work in both Python 2 and 3:

```
>>> re.search('(\d+)\.(\d+)\.(\d+)', "1.2.3").group(1)
'1'
>>> re.search(r'(\d+)\.(\d+)\.(\d+)', "1.2.3").group(1)
'1'
>>> re.search('(\\d+)\.(\\d+)\.(\\d+)', "1.2.3").group(1)
'1'
```

I claim it's weird that the first one works at all, given that the other two do. That's the essence of the warning. Using the second form seems more logical. 

It's not a small change to fix all of them. Although it's straightforward to make the strings into raw strings, there are potential landmines like `re.search('(\d+)\.(\d+)\.(\\d+)', "1.2.3").group(3)`. That's a bad regex as it mixes this confusing Python syntax behavior both ways; making in a raw string breaks the behavior :(

It'd be a worthy project to try to carefully evaluate and clean them up as I think most will be straightforward.

And the rest of the warnings looks like "line break after binary operator" which can be suppressed, and a few lines that are too long.

**Comment by cclauss** [comment]
Let's wait until #22400 is merged before proceeding with this PR.

**Comment by cclauss** [comment]
@srowen Following the merge of #22400 there are just three W605s at the bottom of https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96020/console  Should we __# noqa__ those lines or do you have a slicker solution/  Thanks massively for your persistence.

**Comment by HyukjinKwon** [comment]
shall we fix them here? Looks not many instances

**Comment by cclauss** [comment]
OK...  __# noqa__ works within the flake8 wrapper around PyFlakes but does not work when PyFlakes is called outside of flake8.  Also added r"comment" as you suggested.

**Comment by cclauss** [comment]
OK...  Removed backticks and we are once again testing ./dev/run-tests-jenkins.py

**Comment by srowen** [comment]
Oh, that line isn't the error, although the warning says it is! It's really line 134, which actually has escaped back-ticks. I think you can honestly remove all of these back-ticks.

**Comment by HyukjinKwon** [comment]
> OK... # noqa works within the flake8 wrapper around PyFlakes but does not work when PyFlakes is called outside of flake8. Also added r"comment" as you suggested.

Is this a bug in flake8? `# noqa` should work since I fixed one by `# noqa` - https://github.com/apache/spark/blob/5cdb8a23df6f269d6be0bf3536e9af9e29c4a05f/python/setup.py#L37

**Comment by cclauss** [comment]
Removed the escaped backticks.

@HyukjinKwon Undefined names (F821) is a flake8 error, not a PyCodeStyle error.  That means that PyCodeStyle on its own (this PR) is unable to detect the problem with `__version__`.

**Comment by HyukjinKwon** [comment]
It's weird, `# noqa` should work pep8 (`pycodestyle`) too.

**Comment by HyukjinKwon** [comment]
http://pycodestyle.pycqa.org/en/latest/intro.html

File: dev/run-tests-jenkins.py
Comment (HyukjinKwon) [inline dev/run-tests-jenkins.py]: Can you just backquote here too likewise? Does this fail?

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

File: dev/run-tests-jenkins.py
Comment (srowen) [reply dev/run-tests-jenkins.py]: Yeah maybe but I don't see the value in these anyway. It seems to be trying to quote text, but the quote isn't even meaningful. It's not code.

**Review by srowen — APPROVED** [review]

**Comment by cclauss** [comment]
@holdenk Can you please do me a big favor and create a separate PR that makes sure that we are doing at least __flake8 . --count --select=E901,E999,F821,F822,F823 --show-source --statistics__ on the whole codebase on Python 2 and Python 3?  When I create PRs on this project, I make so many mistakes that it takes pages of conversation and weeks of testing to get them merged.  I just want to make sure that the project never backslides on F821s because it has taken [a lot of effort](https://github.com/apache/spark/pulls?q=is%3Apr+author%3Acclauss+is%3Aclosed) to get us clean.  Thanks.

**Comment by srowen** [comment]
Merged to master/2.4

**Comment by cloud-fan** [comment]
Note that, RC1 was cut before merging this PR, which means, this patch is not available in 2.4.0. I hit some problems running the release scripts and spent quite a lot of time to fix them, so the final vote is several days behind the RC1 tag creation.

@srowen please advice if we should
1. fail the RC1 to include this patch
2. do nothing and release it with 2.4.1
3. revert it from 2.4 since it's an upgrade.

Thanks!

**Comment by srowen** [comment]
Yeah I noticed that. I think we should leave it, and, if somehow RC1 passes, we'll mark this as fixed for a later release.

**Pull request closed without merging**
